webcompat/webhooks/model.py

```diff
@@ -0,0 +1,185 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""WebCompat Issue Model for webhooks."""
+
+from dataclasses import dataclass
+from dataclasses import field
+from typing import Any
+from typing import Dict
+from typing import List
+
+from requests.exceptions import HTTPError
+
+from webcompat import app
+from webcompat.webhooks.helpers import extract_metadata
+from webcompat.webhooks.helpers import get_issue_labels
+from webcompat.webhooks.helpers import make_request
+from webcompat.webhooks.helpers import prepare_rejected_issue
+
+PUBLIC_REPO = app.config['ISSUES_REPO_URI']
+PRIVATE_REPO = app.config['PRIVATE_REPO_URI']
+
+
+@dataclass
+class WebHookIssue:
+    """WebCompat Issue Model for WebHook consumption"""
+    action: str
+    body: str
+    domain: str
+    number: int
+    public_url: str
+    repository_url: str
+    state: str
+    title: str
+    original_labels: List[str]
+    milestone: str
+    milestoned_with: str
+
+    @classmethod
+    def from_dict(cls, payload):
+        """Class method to allow instantiation from a GitHub response dict."""
+        # Extract the title and the body
+        issue = payload.get('issue')
+        full_title = issue.get('title', 'Weird_Title - Inspect')
+        labels = issue.get('labels', [])
+        issue_body = issue.get('body')
+        domain = full_title.partition(' ')[0]
+        public_url = extract_metadata(issue_body).get('public_url', '').strip()
+        original_labels = [label['name'] for label in labels]
+        # webhook with a milestone already set
+        milestone = ''
+        if issue.get('milestone'):
```

Comment:
If the `issue` key can be missing from `payload` you'll get an `AttributeError: 'NoneType' object has no attribute 'get'` at that point. That is probably better to break at line 46 with a direct access `payload['issue']` and the more explicit error: `KeyError: 'issue'`

Reply:
I guess the test here would be what is happening if we get a payload which is not the expected one or no payload because GitHub went down or something on the network broke. 

this is tied to https://github.com/webcompat/webcompat.com/blob/77c102721ce05d74c9372a35b14884b62ab45623/webcompat/webhooks/__init__.py#L33-L37

we know already that's a webhook issue event payload. So that would be really the case where `'X-GitHub-Event'` and the payload contradict themselves. so a try would just send back `ooops` instead of a 500.

Reply:
Re-reading, we can add some error recovery if `issue` is missing from `payload`, but yes, it would imply that GitHub is totally broken.

We do need to check for `milestone` on the `issue`, however, because it's possible for a payload to not have a `milestone` property on `issue` (or at the root level, in the instanced of a `milestoned` webhook type.

Reply:
OK, I've made the change in https://github.com/webcompat/webcompat.com/pull/3415/commits/3e5eadcd330400487e778093345bc68cfee6cb22. I wrote a test, but it doesn't provide much value -- it just asserts that if you pass in garbage, you get a KeyError. It's possible that GitHub does weird stuff one day, but we'll have bigger problems if so (everything will fail, not just this webhook). But the KeyError will provide better clues in logs, so that's nice.